### ISSUE TITLE
Error Compiling Nikon-Patch

### ISSUE THREAD

**Issue by luizzod** [simeonpilgrim/nikon-firmware-tools#87]
I've been trying to set up the nikon patcher on my computer for personal use. I have a D3200 that I'm trying to use as a webcam. There is the website however it is currently unavailable. I know some C++ so I guess why not try to get it to run locally if possible.

I installed node.js (npm) and emscripten to build nikon-patcher-js. 
Then I switch to \Nikon-Patch-JS\src\nikon_patch\

My command:
`emcc nikon_patch.c -s WASM=1 -o nikon_patch.html`
 
It starts linking the libraries however after wasm\libsockets.a it fails. 
wasm-id.exe sends the error "initial memory too small". 

The command line:

`C:\Users\barrera\Documents\Dev\nikon-firmware-tools-master\Nikon-Patch-JS\src\nikon_patch>emcc nikon_patch.c -s WASM=1 -o nikon_patch.html
cache:INFO: generating system asset: is_vanilla.txt... (this will be cached in "C:\Users\barrera\.emscripten_cache\is_vanilla.txt" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libcompiler_rt.a... (this will be cached in "C:\Users\barrera\.emscripten_cache\wasm\libcompiler_rt.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc-wasm.a... (this will be cached in "C:\Users\barrera\.emscripten_cache\wasm\libc-wasm.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libdlmalloc.a... (this will be cached in "C:\Users\barrera\.emscripten_cache\wasm\libdlmalloc.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libpthread_stub.a... (this will be cached in "C:\Users\barrera\.emscripten_cache\wasm\libpthread_stub.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc_rt_wasm.a... (this will be cached in "C:\Users\barrera\.emscripten_cache\wasm\libc_rt_wasm.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libsockets.a... (this will be cached in "C:\Users\barrera\.emscripten_cache\wasm\libsockets.a" for subsequent builds)
cache:INFO:  - ok
wasm-ld: error: initial memory too small, 110108816 bytes needed
wasm-ld: error: C:\Users\barrera\AppData\Local\Temp\emscripten_temp_4y_cnv1s\nikon_patch_0.o: undefined symbol: patches_count
wasm-ld: error: C:\Users\barrera\AppData\Local\Temp\emscripten_temp_4y_cnv1s\nikon_patch_0.o: undefined symbol: patches
wasm-ld: error: C:\Users\barrera\AppData\Local\Temp\emscripten_temp_4y_cnv1s\nikon_patch_0.o: undefined symbol: patches
wasm-ld: error: C:\Users\barrera\AppData\Local\Temp\emscripten_temp_4y_cnv1s\nikon_patch_0.o: undefined symbol: patches_count
wasm-ld: error: C:\Users\barrera\AppData\Local\Temp\emscripten_temp_4y_cnv1s\nikon_patch_0.o: undefined symbol: patches
emcc: error: 'C:/Users/barrera/Documents/Dev/emsdk-master/upstream/bin\wasm-ld.exe -o C:\Users\barrera\AppData\Local\Temp\emscripten_temp_4y_cnv1s\nikon_patch.wasm C:\Users\barrera\AppData\Local\Temp\emscripten_temp_4y_cnv1s\nikon_patch_0.o -LC:\Users\barrera\Documents\Dev\emsdk-master\upstream\emscripten\system\local\lib -LC:\Users\barrera\Documents\Dev\emsdk-master\upstream\emscripten\system\lib -LC:\Users\barrera\.emscripten_cache\wasm C:\Users\barrera\.emscripten_cache\wasm\libc.a C:\Users\barrera\.emscripten_cache\wasm\libcompiler_rt.a C:\Users\barrera\.emscripten_cache\wasm\libc-wasm.a C:\Users\barrera\.emscripten_cache\wasm\libdlmalloc.a C:\Users\barrera\.emscripten_cache\wasm\libpthread_stub.a C:\Users\barrera\.emscripten_cache\wasm\libc_rt_wasm.a C:\Users\barrera\.emscripten_cache\wasm\libsockets.a -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined --import-memory --import-table --strip-debug --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export fflush --export __errno_location -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024' failed (1)`

This is supposed to show me a website just like yours right?

Thanks for the great work.

**Comment by simeonpilgrim** [comment]
build.cmd has the command I use which is:

`..\..\..\..\emsdk\emscripten\incoming\emcc nikon_patch.c patches.c md5.c md5driver.c xor.c -o nikon_patch.js -s WASM=1 -O3 -s TOTAL_MEMORY=115343360 -s EXTRA_EXPORTED_RUNTIME_METHODS="['getValue']"
`

**Comment by luizzod** [comment]
Thanks for the fast response. 

I was able to get the page running with npm but the webpage stays at "Waiting for WebAssembley code to complete loading.... "

**Comment by simeonpilgrim** [comment]
yip, you need to copy the output into the react-app project path, which the cp.cmd has this in it.

`COPY /Y nikon_patch.js ..\..\public\
COPY /Y nikon_patch.wasm ..\..\public\
`

**Comment by luizzod** [comment]
Yes I did that, and double checked however no change in the webpage. 

`C:\Users\barrera\Documents\Dev\nikon-firmware-tools-master\Nikon-Patch-JS\public>dir
 Volume in drive C has no label.
 Volume Serial Number is BCD8-014B

 Directory of C:\Users\barrera\Documents\Dev\nikon-firmware-tools-master\Nikon-Patch-JS\public

04/22/2020  10:44 PM    <DIR>          .
04/22/2020  10:44 PM    <DIR>          ..
04/05/2020  04:21 PM            16,433 changelog.html
04/05/2020  04:21 PM             1,150 favicon.ico
04/05/2020  04:21 PM             5,809 index.html
04/05/2020  04:21 PM               317 manifest.json
04/22/2020  10:42 PM            17,897 nikon_patch.js
04/22/2020  10:42 PM            69,372 nikon_patch.wasm
               6 File(s)        110,978 bytes
               2 Dir(s)  31,640,088,576 bytes free`

Right now I'm gonna try using another older browser around the time you deployed this.

**Comment by simeonpilgrim** [comment]
while developing, I would `build.cmd` and then `cp.cmd` and then hit F5 on the web browser.. and would see the updates..

**Comment by luizzod** [comment]
The console gave me this error:
```
Loading failed for the <script> with source “http://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js”. nikon-patch:22:1

The script from “http://localhost:3000/nikon_patch.js” was loaded even though its MIME type (“text/html”) is not a valid JavaScript MIME type.

nikon-patch
SyntaxError: expected expression, got '<'
nikon_patch.js:1

ReferenceError: Module is not defined
nikon-patch:49:9

    <anonymous> http://localhost:3000/nikon-patch:49
[HMR] Waiting for update signal from WDS...`
```

It points to this part in the nikon_patch.js:
```
<script type="text/javascript" src="nikon_patch.js"></script>

    <script type="text/javascript">

        Module['onRuntimeInitialized'] = function() { 

            var loading = document.getElementById('webasmloading');

            loading.style.display = 'none';

            var rootwrap = document.getElementById('rootwrap');

            rootwrap.style.display = 'block';

            var ret = _GeneratePatchMatrix();

            if(ret>0){

                var outptr = _getOutFilePtr();

                let s = "";

                for (let i = 0; i < ret; ++i){

                    s += String.fromCharCode(Module.getValue(outptr+i));

                }

                var featurelist = document.getElementById('featurelist');

                featurelist.innerHTML = s;

            }

        };

    </script>
```

**Comment by simeonpilgrim** [comment]
ah I see that problem.. I was testing on my work pc when bumping the packages versions, which I didn't have web-assembly compiler installed on, so didn't think anything of that not loading.. as the reset of the page loaded.

**Comment by luizzod** [comment]
It worked with that version. The bitrate has been improved on the D3200 however I haven't tried beta. 
Thank you so much for your work. Im still kinda hoping for getting the time-limit off as that would be ideal but this is good.

**Comment by simeonpilgrim** [comment]
fixed relative pathing now in [898b792]

**Comment by simeonpilgrim** [comment]
thanks for using the tools locally and thus finding this bug.
